dmoj/judge.py

```diff
@@ -175,41 +179,202 @@ def begin_grading(
         blocking=False,
     ):
         self.current_submission_id = id
+        sub_judge_thread = threading.Thread(
+            target=self._run_grading_thread,
+            args=(id, problem_id, language, source, time_limit, memory_limit, short_circuit, meta, report),
+            daemon=True,
+        )
+        sub_judge_thread.start()
 
-        def grading_cleanup_wrapper():
-            report(
-                ansi_style(
-                    'Start grading #ansi[%s](yellow)/#ansi[%s](green|bold) in %s...' % (problem_id, id, language)
-                )
-            )
+        # Submission may have been killed already, but block if it hasn't been.
+        if blocking and sub_judge_thread is not None:
+            sub_judge_thread.join()
+
+    def log_internal_error(self, exc=None, message=None):
+        if not message:
+            # If exc is exists, raise it so that sys.exc_info() is populated with its data
+            if exc:
+                try:
+                    raise exc
+                except:  # noqa: E722, we want to catch everything
```

Comment:
This existed before, but technically this is a race because the user could Ctrl-C just before `raise exc` and `sys.exc_info()` would be `KeyboardInterrupt`.

Reply:
Fixed.